dali/operators/reader/loader/video/video_file.cc

```
@@ -0,0 +1,232 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/reader/loader/video/video_file.h"
+#include "dali/core/error_handling.h"
+
+namespace dali {
+
+void VideoFileCPU::InitAvState() {
+  codec_ctx_ = avcodec_alloc_context3(codec_);
+  DALI_ENFORCE(codec_ctx_, "Could not create av codec context");
+
+  DALI_ENFORCE(
+    avcodec_parameters_to_context(codec_ctx_, codec_params_) >= 0,
+    "Could not fill the codec based on parameters");
+
+  DALI_ENFORCE(
+    avcodec_open2(codec_ctx_, codec_, nullptr) == 0,
+    "Could not initialize codec context");
+
+  frame_ = av_frame_alloc();
+  DALI_ENFORCE(frame_, "Could not allocate the av frame");
+
+  packet_ = av_packet_alloc();
+  DALI_ENFORCE(packet_, "Could not allocate av packet");
+}
+
+void VideoFileCPU::FindVideoStream() {
+  for (size_t i = 0; i < ctx_->nb_streams; ++i) {
+    auto stream = ctx_->streams[i];
+    codec_params_ = ctx_->streams[i]->codecpar;
+    codec_ = avcodec_find_decoder(codec_params_->codec_id);
+
+    if (codec_ == nullptr) {
+      continue;
+    }
+
+    if (codec_->type == AVMEDIA_TYPE_VIDEO) {
+        stream_id_ = i;
+        return;
+    }
+  }
+
+  DALI_FAIL(make_string("Could not find a valid video stream in file ", filename_));
+}
+
+VideoFileCPU::VideoFileCPU(std::string &filename) : 
+  filename_(filename) {
+  ctx_ = avformat_alloc_context();
+  DALI_ENFORCE(ctx_, "Could not create avformat context");
+
+  DALI_ENFORCE(
+    avformat_open_input(&ctx_, filename.c_str(), nullptr, nullptr) == 0,
+    make_string("Failed to open video file at path ", filename));
+
+  FindVideoStream();
+  InitAvState();
+  BuildIndex();
+}
+
+void VideoFileCPU::BuildIndex() {
+  // TODO(awolant): Optimize this function for: 
+  //  - CFR
+  //  - index present in the header
+
+  int last_keyframe = -1;
+  while (ReadRegularFrame(nullptr, false)) {
+    IndexEntry entry;
+    entry.is_keyframe = frame_->key_frame;
```

Maybe frame should be returned by ReadRegularFrame and not silently passed by frame_?

---

As I see it, `AVFrame` isn't something that makes sense to return. It is mostly part of the ffmpeg decoder. It is basically a reusable container that `libav` keeps for us and fills with data, when we ask it to. To make this design more clear, I extracted all of the pointers to various `ffmpeg` structures and combined them in `AvState` struct. Now, it is clear, that with various functions from `libav` we just manipulate this state and we can copy data from it, when we know it is ready.